ws/api/endpoints.go

```diff
@@ -0,0 +1,173 @@
+// Copyright (c) Mainflux
+// SPDX-Licence-Identifier: Apache-2.0
+
+package api
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/go-zoo/bone"
+	"github.com/gorilla/websocket"
+	"github.com/mainflux/mainflux/pkg/errors"
+	"github.com/mainflux/mainflux/pkg/messaging"
+	"github.com/mainflux/mainflux/ws"
+)
+
+var (
+	channelPartRegExp = regexp.MustCompile(`^/channels/([\w\-]+)/messages(/[^?]*)?(\?.*)?$`)
+)
+
+func handshake(svc ws.Service) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		req, err := decodeRequest(r)
+		if err != nil {
+			switch err {
+			case errUnauthorizedAccess:
+				w.WriteHeader(http.StatusForbidden)
+				return
+			default:
+				logger.Warn(fmt.Sprintf("Failed to authorize: %s", err.Error()))
+				w.WriteHeader(http.StatusServiceUnavailable)
+				return
+			}
+		}
+
+		// Upgrade to a new ws connection.
+		conn, err := upgrader.Upgrade(w, r, nil)
+		if err != nil {
+			logger.Warn(fmt.Sprintf("Failed to upgrade connection to websocket: %s", err.Error()))
+			return
+		}
+
+		// Configure the closing of the ws connection
+		conn.SetCloseHandler(func(code int, text string) error {
+			return svc.Unsubscribe(context.Background(), req.thingKey, req.chanID, req.subtopic)
+		})
```

Comment:
As per `SetCloseHandler` documentation:

>  Applications should only set a close handler when the application must perform some action before sending a close message back to the peer.

Can you think of other ways of unsubscribing (maybe when you close the `msgs` channel)?